soe/basic.go

```diff
@@ -0,0 +1,73 @@
+package soe
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/hamba/avro/v2"
+)
+
+// Codec marshals values to/from bytes, with the Avro binary payload wrapped in
+// an SOE frame containing the writer schema fingerprint.
+type Codec struct {
+	api    avro.API
+	schema avro.Schema
+	header []byte
+}
+
+// NewCodec creates a new Codec for a Schema and the default config.
+func NewCodec(schema avro.Schema) (*Codec, error) {
+	return NewCodecWithAPI(schema, avro.DefaultConfig)
+}
+
+// NewCodecWithAPI creates a new Codec for a Schema and an API.
+func NewCodecWithAPI(schema avro.Schema, api avro.API) (*Codec, error) {
+	// Precompute SOE header
+	header, err := BuildHeader(schema)
+	if err != nil {
+		return nil, err
+	}
+	return &Codec{
+		schema: schema,
+		api:    api,
+		header: header,
+	}, nil
+}
+
+// Encode marshals a value to SOE-encoded Avro binary.
+func (c *Codec) Encode(v any) ([]byte, error) {
+	data, err := c.api.Marshal(c.schema, v)
+	if err != nil {
+		return nil, err
+	}
+	return append(c.header, data...), nil
+}
+
+// Decode unmarshals a value from SOE-encoded Avro binary, and fails if
+// the schema fingerprint doesn't match the held schema.
+func (c *Codec) Decode(data []byte, v any) error {
```

Comment:
It would be really handy to have an interface for the decoders;
```
type Decoder interface {
	Decode(data []byte, v any)
	DecodeUnverified(data []byte, v any)
}
```
because that would cover both `Codec` and `DynamicDecoder`, so the more complicated dynamic decoder can be swapped out in tests. Unfortunately I don't think it's possible to fold the `TypedCodec` into the same interface.

Reply:
It is unconventional to add this here, but rather it would be better for the user to define it where it is used if it is needed, IMO.